src/lib/process.ts

```diff
@@ -244,3 +235,166 @@ export function clearConsole() {
+    isAppProject = typeof require('./package.json')?.dependencies?.[tool.depName] === 'string'
+  } catch (e) {}
+
+  if (!isAppProject) {
+    return {
+      nodeProject: true,
+      toolProject: false,
+      toolCurrentlyPresentInNodeModules: false,
+      runningLocalBin: false,
+      paths: {
+        thisProcessBinDir,
+        projectDir,
+        projectBinDir,
+      },
+    }
+  }
+
+  let toolCurrentlyPresentInNodeModules = fs.existsSync(path.join(projectNodeModulesDir, tool.depName))
+
+  if (!toolCurrentlyPresentInNodeModules) {
+    return {
+      nodeProject: true,
+      toolProject: true,
+      toolCurrentlyPresentInNodeModules: false,
+      runningLocalBin: false,
+      paths: {
+        thisProcessBinDir,
+        projectDir,
+        projectBinDir,
+      },
+    }
+  }
+
+  if (thisProcessBinDir !== projectBinDir) {
+    return {
+      nodeProject: true,
+      toolProject: true,
+      toolCurrentlyPresentInNodeModules: true,
+      runningLocalBin: false,
+      paths: {
+        thisProcessBinDir,
+        projectDir,
+        projectBinDir,
+      },
+    }
+  }
+
+  return {
+    nodeProject: true,
+    toolProject: true,
+    toolCurrentlyPresentInNodeModules: true,
+    runningLocalBin: true,
+    paths: {
+      thisProcessBinDir,
+      projectDir,
+      projectBinDir,
+    },
+  }
+}
+
+/**
+ * Handoff execution from a global to local version of a package.
+ *
+ * If the givne global module path is not a real node package (defined as being
+ * unable to locate its package.json file) then an error will be thrown.
+ *
+ * An environment variable called `GLOBAL_LOCAL_HANDOFF` will be set to
+ * `"true"`. Use this to short-circuit startup logic.
+ */
+export function globalLocalHandoff(input: { localPackageDir: string; globalPackageFilename: string }) {
+  if (process.env.GLOBAL_LOCAL_HANDOFF) {
+    console.warn('warning: multiple calls to `globalLocalHandoff`, this should not happen.')
+  }
+
+  process.env.GLOBAL_LOCAL_HANDOFF = 'true'
+
+  const globalProjectDir = findFilenameProjectDir(input.globalPackageFilename)
+
+  if (!globalProjectDir) {
+    throw new Error(
+      `Could not perform handoff to local package version becuase the given global package does not appear to actually be a package:\n\n${input.globalPackageFilename}`
+    )
+  }
+
+  require(path.join(input.localPackageDir, path.relative(globalProjectDir, input.globalPackageFilename)))
+}
+
+/**
+ * Given a module file path find the path to the project dir containing it
+ * defined as the first ancestor dir with a package.json.
+ *
+ * @return
+ *
+ * `null` if no ancestor dir has a package.json file, otherwise the path to
+ * project dir.
+ */
+export function findFilenameProjectDir(filename: string): null | string {
+  let dir = path.dirname(filename)
+  while (true) {
+    if (fs.existsSync(path.join(dir, 'package.json'))) {
+      return dir
+    }
+    if (dir === '') {
+      return null
+    }
+    dir = path.dirname(dir)
+  }
+}
```

Comment:
What's the difference between that function and `findFileRecursivelyUpwardSync('package.json', { cwd: path.dirname('filename') })` ?